[lldb] Extend FindTypes to optionally search by mangled type name

augusto2112 opened the pull request:
Swift types have mangled type names. This adds functionality to look up those types through the FindTypes API by searching for the mangled type name instead of the regular name.

augusto2112 commented:
This depends on https://github.com/llvm/llvm-project/pull/113006

Michael137 commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFDIE.cpp:
```diff
@@ -434,6 +442,19 @@ static void GetTypeLookupContextImpl(DWARFDIE die,
+
+    // If there is no name, then there is no need to look anything up for this
+    // DIE.
+       const char *name = die.GetName();
```
why is this required now, whereas it wasn't before?

Michael137 commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFDIE.cpp:
```diff
@@ -453,7 +474,7 @@ static void GetTypeLookupContextImpl(DWARFDIE die,
     case DW_TAG_typedef:
       push_ctx(CompilerContextKind::Typedef, die.GetName());
-      break;
+   break;
```
something weird happened to the format here

Michael137 commented on lldb/source/Plugins/SymbolFile/CTF/SymbolFileCTF.cpp:
```diff
@@ -1032,10 +1032,16 @@ void SymbolFileCTF::FindTypes(const lldb_private::TypeQuery &match,
-      results.InsertUnique(type_sp);
-      if (results.Done(match))
-        return;
+    if (type_sp) {
```
If you're already rewriting this, lets do:
```
if (!type_sp)
  continue;
```
(same applies to the `type_name == name` condition below)

Michael137 commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFDIE.cpp:
```diff
@@ -434,6 +442,19 @@ static void GetTypeLookupContextImpl(DWARFDIE die,
     };
+
+    // Since mangled names are unique there's no need to build an entire context.
+    if (use_mangled_name) {
```
If all that `use_mangled_name` is doing is to create a vector with 1 entry, which is the mangled name, can we just do so at the callsite of this function, instead of embedding it here?

Michael137 commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp:
```diff
@@ -2762,7 +2762,7 @@ void SymbolFileDWARF::FindTypes(const TypeQuery &query, TypeResults &results) {
       die_context = die.GetDeclContext();
     else
-      die_context = die.GetTypeLookupContext();
+      die_context = die.GetTypeLookupContext(query.GetSearchByMangledName());
```
To avoid mixing mangled typenames into the `CompilerDeclContext` logic, we could just compare mangled names here if `GetSearchByMangledName()`, instead of decl context?

augusto2112 committed:
[lldb] Extend FindTypes to optionally search by mangled type name

Swift types have mangled type names. This adds functionality to look
up those types through the FindTypes API by searching for the mangled
type name instead of the regular name.

clayborg reviewed (commented):
Typenames don't often get put into the acclerator tables as mangled names for C/C++/ObjC. Functions do, but not typenames. Is there a case where the debug info contains a mangled name for a type? 

I know swift does use mangled names for types though as that is how these types are extracted from the Swift AST data blob by the Swift compiler. 

So I see no problem with this patch. If this is for Swift then we need to make sure there is a test for this in the Swift.org LLDB. Not sure if we can write a test using C/C++/ObjC or and language that comes from Clang ASTs.

Michael137 commented on lldb/include/lldb/Symbol/Type.h:
```diff
@@ -300,6 +303,12 @@ class TypeQuery {
+  /// searched as a mangled name.
+  bool GetSearchByMangledName() const {
+    return (m_options & e_search_by_mangled_name) != 0;
+  }
```
Is there a setter that only lives on the Apple fork then? Don't think it'd be too outrageous to add both the getter/setter here as we do for all the other options (even if we only have coverage for it on the Apple fork i guess)

Michael137 reviewed (approved):
Seems reasonable to me

Yea not entirely sure how to test this either. As Greg points out, type mangling isn't really a thing LLDB cared about for C++. Maybe if we had this option exposed somehow (e.g., through `lldb-test`) we could come up with a DWARF yaml that has a type with a linkage name in the accelerator tables? Then we could test it that way. But if that's too much hassle and we have a test on the Apple fork that exercises this API maybe we're ok

Michael137 commented:
Yea using the new `TypeQuery` option in `lldb-test` seems doable. Whether `obj2yaml`/`yaml2obj` supports `.debug_names`, i'm not sure, haven't tried

Michael137 commented on lldb/source/Plugins/SymbolFile/CTF/SymbolFileCTF.cpp:
```diff
@@ -1032,11 +1032,17 @@ void SymbolFileCTF::FindTypes(const lldb_private::TypeQuery &match,
+      continue;
+    results.InsertUnique(type_sp);
+    if (results.Done(match))
+      return;
```
Also, do we need to implement this for new lookup option for `CTF` and `PDB`? Given it will probably be always unused (e.g., CTF is only used with `C` IIUC), maybe lets limit this change to just DWARF for now?

augusto2112 replied:
Sure, I implemented there for completeness.

clayborg commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp:
```diff
@@ -2758,6 +2758,15 @@ void SymbolFileDWARF::FindTypes(const TypeQuery &query, TypeResults &results) {
 
+    // Since mangled names are unique, we only need to check if the names are
+    // the same.
+    if (query.GetSearchByMangledName()) {
+      if (die.GetMangledName() == query.GetTypeBasename().GetStringRef())
+        if (Type *matching_type = ResolveType(die, true, true))
+          results.InsertUnique(matching_type->shared_from_this());
+      return !results.Done(query); // Keep iterating if we aren't done.
+    }
```
This logic needs doesn't need to check if results are done if we didn't add anything to the results. Maybe this would be more clear?:
```
    if (query.GetSearchByMangledName()) {
      if (die.GetMangledName() != query.GetTypeBasename().GetStringRef()) 
        return true; // Keep iterating over index types, mangled name mismatch.
      if (Type *matching_type = ResolveType(die, true, true)) {
        results.InsertUnique(matching_type->shared_from_this());
        return !results.Done(query); // Keep iterating if we aren't done.
      }
      return true; // Keep iterating over index types, weren't able to resolve this type
    }
```
One other issue is that `DWARFDie::GetMangledName()` will return the `DW_AT_name` if there is no mangled name:
```
const char *DWARFDIE::GetMangledName() const {
  if (IsValid())
    return m_die->GetMangledName(m_cu);
  else
    return nullptr;
}
```
But `const char *DWARFDebugInfoEntry::GetMangledName(const DWARFUnit *cu, bool substitute_name_allowed)` has a boolean option `substitute_name_allowed` that defaults to true. So we should probably also change:
```
const char *DWARFDIE::GetMangledName() const;
```
To have that same option with a default value:
```
const char *DWARFDIE::GetMangledName(bool substitute_name_allowed = true) const
```
And then change your code to pass in `false` for the `substitute_name_allowed`

augusto2112 replied:
Good catch! Thanks!

augusto2112 commented:
@Michael137 I wasn't able to run `yaml2obj` to convert back an object file that contains debug info, I'm not sure if the issue is that `yaml2obj` doesn't support something that's encoded in the debug info, or some bug with MachO.

This is the error:

```
error: wrote too much data somewhere, section offsets in section __debug_str_offs__DWARF for segment __DWARF don't line up: [cursor=0x41d1e0], [fileStart=0x41ccdc], [sectionOffset=0x4f2]
```

Michael137 commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp:
```diff
@@ -2758,6 +2758,20 @@ void SymbolFileDWARF::FindTypes(const TypeQuery &query, TypeResults &results) {
+    // Since mangled names are unique, we only need to check if the names are
+    // the same.
+    if (query.GetSearchByMangledName()) {
+      if (die.GetMangledName(/*substitute_name_allowed=*/false) !=
```
Even with substitutions, that would only cause problems if the `DW_AT_name` somehow contained a the mangled name, right? Even then, that might not actually be a problem.

But I guess being more explicit about really only checking the linkage name seems fine

Michael137 commented:
> @Michael137 I wasn't able to run `yaml2obj` to convert back an object file that contains debug info, I'm not sure if the issue is that `yaml2obj` doesn't support something that's encoded in the debug info, or some bug with MachO.
> 
> This is the error:
> 
> ```
> error: wrote too much data somewhere, section offsets in section __debug_str_offs__DWARF for segment __DWARF don't line up: [cursor=0x41d1e0], [fileStart=0x41ccdc], [sectionOffset=0x4f2]
> ```

Hmm yea I wouldn't be surprised if `yaml2obj` doesn't properly handle this. Could give it a shot using ELF (I think that has better support). But if that doesn't work either I wouldn't bother

adrian-prantl reviewed (commented):
Can you add a test (perhaps using lldb-test and a .ll or .s file)?

adrian-prantl commented on lldb/test/Shell/SymbolFile/DWARF/debug-types-mangled-name.ll:
```diff
@@ -0,0 +1,63 @@
+  ret i32 0, !dbg !26
+}
+
+attributes #0 = { noinline nounwind optnone uwtable "frame-pointer"="non-leaf" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="generic" "target-features"="+fp-armv8,+neon,+v8a,-fmv" }
```
Can you try to remove most of the attributes?

augusto2112 committed:
remote attributes from test
